core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -909,6 +774,237 @@ private void initialize() {
         }
     }
 
+    private Map<Long, RecordState> fetchRecordStateMapForAcknowledgementBatch(AcknowledgementBatch batch) {
+        // Client can either send a single entry in acknowledgeTypes which represents the state
+        // of the complete batch or can send individual offsets state. Construct a map with record state
+        // for each offset in the batch, if single acknowledge type is sent, the map will have only one entry.
+        Map<Long, RecordState> recordStateMap = new HashMap<>();
+        for (int index = 0; index < batch.acknowledgeTypes.size(); index++) {
+            recordStateMap.put(batch.firstOffset() + index,
```

Comment:
Why are we doing batch.firstOffset()+index instead of just batch.firstOffset()

Reply:
Becuse the index starts from 0

Reply:
ahh okay, I misread the for loop, I thought we were iterating it over multiple batches of acknowledgements. Thanks for clearing it.